presto-function-namespace-managers/src/main/java/com/facebook/presto/functionNamespace/rest/RestSqlFunctionExecutor.java

```diff
@@ -0,0 +1,227 @@
+    private static PagesSerde pageSerde;
+    private HttpClient httpClient;
+    private final RestBasedFunctionNamespaceManagerConfig restBasedFunctionNamespaceManagerConfig;
+    public static final String PRESTO_PAGES = "application/X-presto-pages";
+
+    @Inject
+    public RestSqlFunctionExecutor(RestBasedFunctionNamespaceManagerConfig restBasedFunctionNamespaceManagerConfig, @ForRestServer HttpClient httpClient)
+    {
+        this.restBasedFunctionNamespaceManagerConfig = requireNonNull(restBasedFunctionNamespaceManagerConfig, "restBasedFunctionNamespaceManagerConfig is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient is null");
+    }
+
+    @Override
+    public FunctionImplementationType getImplementationType()
+    {
+        return REST;
+    }
+
+    @Override
+    public void setBlockEncodingSerde(BlockEncodingSerde blockEncodingSerde)
+    {
+        checkState(this.blockEncodingSerde == null, "blockEncodingSerde already set");
+        this.blockEncodingSerde = requireNonNull(blockEncodingSerde, "blockEncodingSerde is null");
+        this.pageSerde = new PagesSerde(blockEncodingSerde, Optional.empty(), Optional.empty(), Optional.empty());
+    }
+
+    @Override
+    public CompletableFuture<SqlFunctionResult> executeFunction(
+            String source,
+            RemoteScalarFunctionImplementation functionImplementation,
+            Page input,
+            List<Integer> channels,
+            List<Type> argumentTypes,
+            Type returnType)
+    {
+        SqlFunctionHandle functionHandle = functionImplementation.getFunctionHandle();
+        SqlFunctionId functionId = functionHandle.getFunctionId();
+        String functionVersion = functionHandle.getVersion();
+        DynamicSliceOutput sliceOutput = new DynamicSliceOutput((int) input.getRetainedSizeInBytes());
+        writeSerializedPage(sliceOutput, pageSerde.serialize(input));
+        try {
+            Request request = preparePost()
+                    .setUri(getExecutionEndpoint(functionId, functionVersion))
+                    .setBodyGenerator(createStaticBodyGenerator(sliceOutput.slice().byteArray()))
+                    .setHeader(CONTENT_TYPE, PRESTO_PAGES)
+                    .setHeader(ACCEPT, PRESTO_PAGES)
+                    .build();
+            HttpClient.HttpResponseFuture<SqlFunctionResult> future = httpClient.executeAsync(request, new SqlFunctionResultResponseHandler());
+            Futures.addCallback(future, new SqlResultFutureCallback(), directExecutor());
+            return toCompletableFuture(future);
+        }
+        catch (Exception e) {
+            return failedFuture(new IllegalStateException("Failed to get function definitions for REST server/ Native worker, " + e.getMessage()));
+        }
+    }
+
+    private URI getExecutionEndpoint(SqlFunctionId functionId, String functionVersion)
+    {
+        if (restBasedFunctionNamespaceManagerConfig.getRestUrl() == null) {
```

Comment:
Because the getter is annotated with @NotNull, I believe this is pre-validated and you can skip this.